### DISCUSSION TITLE
Nextjs Link and Chakra Button

### DISCUSSION THREAD

**Discussion by phifa** [chakra-ui/chakra-ui#3292] [Q&A] [answered]
Hey folks 
Im using next.ja and would like to use a chakra button to link to another internal page. How do I go about doing this?

```         
<Button type="submit">
  Login
</Button>
```

This I know how to do. But how to link to internal Page?

**Comment by chr-ge** [comment, marked as answer]
If you want to use the `Button` component, then you would use `next/router`:
```jsx
import { useRouter } from 'next/router'
import { Button } from '@chakra-ui/react'

const Page = () => {
  const router = useRouter()
  return(
    <Button onClick={() => router.push('/login')}>
      Login
    </Button>
  )
}

export default Page
```

**Reply by phifa** [reply]
Interesting! Thanks for the solution. Would there also be a way to use next/link and make it look like a chakra button?

**Reply by chr-ge** [reply]
You can wrap the `Button` in `NextLink / Link`:

```jsx
import NextLink from 'next/link'

<NextLink href='/login' passHref>
  <Button as="a">Login</Button>
</NextLink>
```

**Reply by phifa** [reply]
ha thats so cool, i thought this would produce invalid html, but it seems to work :)

**Reply by drscottlobo** [reply]
I got an error "a" cannot appear as descendent of "a", but as soon as I removed the as="a" from the button it worked beautifully.

**Comment by mwmcode** [comment]
I had to do the following to make it accessible (for it to get focused on tab)
```jsx
<Button variant="outline" tabIndex={-1} p={0}>
  <Link href="/about">
    <a className="w-full p-2">
      LEARN MORE...
    </a>
  </Link>
</Button>
```
> I think buttons that navigate on click are not friendly from an accessibility perspective
> If I have to use buttons as links, I'd add aria-label to tell screen readers that this is a link to /where/it's/going

**Reply by malashevskyi** [reply]
**`a` tag must not be inside `button`**
![screenshot](https://i.imgur.com/I1chuif.png)

**Reply by mwmcode** [reply]
And vice-verse :\ 

![not](https://user-images.githubusercontent.com/13101565/113428545-2f0a1480-9401-11eb-8859-22ea8da7bf55.png)



Update: [SO answer](https://stackoverflow.com/a/6393863/2535267) so basically children of interactive elements should not be interactive themselves

**Reply by malashevskyi** [reply]
:confused: it is obvious...

**Comment by crushjz** [comment]
You can add `as="a"` on the Chakra Button to transform it.

```jsx
import NextLink from 'next/link'

<NextLink href='/login'>
  <Button as="a">Login</Button>
</NextLink>
```

This will produce a valid HTML anchor.

**Reply by payapula** [reply]
Adding to this, for proper semantics,

```javascript
     <Link href="/login" passHref legacyBehavior>
        <Button as="a">Login</Button>
      </Link>
```

Button as "a" would render anchor tag, and `passHref` on LInk would pass the href down to the Chakra's anchor for proper semantics. Read more here on [passHref in NextJS Link](https://nextjs.org/docs/api-reference/next/link#if-the-child-is-a-custom-component-that-wraps-an-a-tag)

**Reply by andrerpena** [reply]
Payapula's answer is the only right one. It properly renders as:

```
<a class="chakra-button css-8pcd7y" href="/login">Login</a>
```

**Reply by dlhck** [reply]
In newer versions you also need to add the `legacyBehavior` flag the the `NextLink`

**Comment by MincePie** [comment]
I've tried each of these suggestions in a nav button and can't get any of them to direct to my auth page. Has anyone found a solution that works for next.js?

**Reply by payapula** [reply]
All the answers here are indeed for NextJS. 

Can you please reproduce your issue in the below sandbox and post it here? So that we can take a look and guide you. 

https://codesandbox.io/s/chakra-ui-next-js-typescript-kxvyr

**Comment by j-lee8** [comment]
I have a question regarding this. If I hover over the button link, it doesn't show the target URL in the bottom left of the browser window? If I take out the IconButton it does. Very strange!!

```
<NextLink href={`/ideas/edit/${idea.id}`} passHref>
								<IconButton
									color={'cbPrimary.500'}
									_hover={{ color: 'cbPrimary.700' }}
									icon={<PencilIcon />}
									variant={'ghost'}
									aria-label={'Edit car button'}
									title={'Edit this car'}
									size={'xs'}
								/>
							</NextLink>
```

**Reply by payapula** [reply]
With `IconButton` we are rendering a `<button>` 

Hovering on `<a>` tags would display the URL at the left-bottom of the browser. Can you try giving `as` prop to IconButton and render an anchor tag?


```jsx
<NextLink href={`/ideas/edit/${idea.id}`} passHref legacyBehavior>
  <IconButton
          as="a"  <-------- Make this component render <a> instead of the default <button>
	  color={'cbPrimary.500'}
	  _hover={{ color: 'cbPrimary.700' }}
	  icon={<PencilIcon />}
	  variant={'ghost'}
	  aria-label={'Edit car button'}
	  title={'Edit this car'}
	  size={'xs'}
  />
</NextLink>
```
 
 
 **EDIT**: Added `legacyBehavior` prop to preserve old behavior of nextjs. See details in [next 13 breaking changes docs](https://nextjs.org/blog/next-13#breaking-changes)

**Reply by j-lee8** [reply]
Amazing! Thank you :) I'm new to Chakra so still working it out, but thus far I love it. Very easy and flexible!

**Reply by PhilippeSchraepen** [reply]
Can we update this example with the new Link component ([documentation](https://nextjs.org/docs/api-reference/next/link))? 

I managed to get it to work in the same way, with passing the legacyBehaviour tag, but I feel there must be a better way.

**Reply by payapula** [reply]
@PhilippeSchraepen Thanks for pointing this out. It's updated now ✅

**Comment by baga89** [comment]
// Next.js 13

import NextLink from 'next/link';

`<Button as={NextLink} href="/contact">Contact us</Button>`

This will convert `button` to `a` tag, which is better for SEO

`<a class="chakra-button css-6qoqad" href="/contact">Contact us</a>`
